File: backend/jobradar/taxonomy.py
"""Query understanding backed by shared/role_taxonomy.json.

Plain term matching cannot tell "data analyst" from "data scientist" from "data engineer",
because all three share the strongest word in the query. Measured on the live corpus, a
search for "data analyst" returned 115 results of which only 17 contained both words — the
rest were Data Engineers, Financial Analysts and anything else with "data" or "analyst" in
it somewhere.

This module maps a free-text query onto a role family so that ranking can do two things
term matching cannot: recognise that "BI Analyst" is the same job under a different name,
and actively demote the sibling roles that merely share vocabulary.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass, field
from functools import lru_cache
from pathlib import Path
from typing import Any


def shared_dir() -> Path:
    here = Path(__file__).resolve()
    for candidate in here.parents:
        if (candidate / "shared" / "role_taxonomy.json").exists():
            return candidate / "shared"
    raise FileNotFoundError("could not locate shared/ directory")


@lru_cache(maxsize=1)
def taxonomy() -> dict[str, Any]:
    return json.loads((shared_dir() / "role_taxonomy.json").read_text())


@lru_cache(maxsize=1)
def families() -> dict[str, dict[str, Any]]:
    return {f["id"]: f for f in taxonomy()["families"]}


@lru_cache(maxsize=1)
def noise_titles() -> frozenset[str]:
    return frozenset(t.lower() for t in taxonomy().get("noise_titles", []))


@lru_cache(maxsize=1)
def seniority_order() -> list[str]:
    return list(taxonomy().get("seniority_order", []))


@lru_cache(maxsize=1)
def _seniority_phrases() -> list[tuple[str, str]]:
    """(normalised phrase, level), longest first so "entry level" beats "entry"."""
    pairs = [
        (normalise(term), level)
        for level, terms in taxonomy().get("seniority_terms", {}).items()
        for term in terms
        # Roman-numeral and single-letter markers are grade suffixes on a title, not
        # something anybody types into a search box, and "i" would fire on every query.
        if len(term) > 2
    ]
    return sorted(pairs, key=lambda kv: -len(kv[0]))


def detect_query_seniority(query_norm: str) -> str:
    """The experience level a query asks for, or "" when it asks for none.

    Without this, "entry level data scientist", "junior data scientist" and "data
    scientist" were the same search: all three returned the identical 27 results led by
    Sr and Principal roles, because the seniority word matched no title and was simply
    carried along as dead weight in the term list.
    """
    for phrase, level in _seniority_phrases():
        if re.search(rf"(?<![a-z0-9]){re.escape(phrase)}(?![a-z0-9])", query_norm):
            return level
    return ""


def seniority_distance(asked: str, offered: str) -> int:
    """How many rungs apart two levels are on the ladder, or 0 when either is unknown."""
    order = seniority_order()
    if asked not in order or offered not in order:
        return 0
    return abs(order.index(asked) - order.index(offered))


def normalise(text: str) -> str:
    """Lowercase, expand the abbreviations these titles are riddled with, collapse space."""
    t = f" {text.lower()} "
    t = re.sub(r"[^a-z0-9+#/&.\- ]+", " ", t)
    # Expand before collapsing, so "Sr." and "Sr" both become "senior".
    for pattern, repl in (
        (r"\bsr\b\.?", "senior"),
        (r"\bjr\b\.?", "junior"),
        (r"\bmgr\b\.?", "manager"),
        (r"\beng\b\.?", "engineer"),
        (r"\bdev\b\.?", "developer"),
        (r"\banalytics\b", "analytics"),
        (r"\bml\b", "machine learning"),
        (r"\bai\b", "artificial intelligence"),
        (r"\bbi\b", "business intelligence"),
        (r"\bds\b", "data science"),
        (r"\bswe\b", "software engineer"),
        (r"\bsde\b", "software engineer"),
        (r"\bpm\b", "product manager"),
    ):
        t = re.sub(pattern, repl, t)
    return re.sub(r"\s+", " ", t).strip()


@dataclass
class Intent:
    """What a query is actually asking for."""

    raw: str
    normalised: str
    terms: list[str] = field(default_factory=list)
    family_id: str = ""
    # Phrases that mean the same role, used as positive title evidence.
    aliases: list[str] = field(default_factory=list)
    # Skills that corroborate the family when the title is vague.
    skills: list[str] = field(default_factory=list)
    # Canonical phrases of sibling families — a title matching one of these is probably a
    # different job, however many query words it happens to contain.
    rivals: list[str] = field(default_factory=list)
    # The experience level the query asked for, "" when it asked for none.
    seniority: str = ""

    @property
    def has_family(self) -> bool:
        return bool(self.family_id)


def _all_phrases(fam: dict[str, Any]) -> list[str]:
    return [normalise(p) for p in (fam.get("canonical", []) + fam.get("aliases", []))]


def detect_family(query_norm: str) -> str:
    """Pick the family a query is asking for, or "" when it names none.

    Longest phrase wins: "machine learning engineer" must beat "machine learning" so an ML
    Engineer search does not resolve to Data Scientist.
    """
    best_id, best_len = "", 0
    for fid, fam in families().items():
        for phrase in _all_phrases(fam):
            if phrase and phrase in query_norm and len(phrase) > best_len:
                best_id, best_len = fid, len(phrase)
    return best_id


def understand(query: str) -> Intent:
    """Turn a free-text query into an Intent."""
    norm = normalise(query)
    intent = Intent(
        raw=query,
        normalised=norm,
        terms=[t for t in norm.split() if len(t) > 1],
        seniority=detect_query_seniority(norm),
    )

    fid = detect_family(norm)
    if not fid:
        return intent

    fam = families()[fid]
    intent.family_id = fid
    intent.aliases = _all_phrases(fam)
    intent.skills = [normalise(s) for s in fam.get("skills", [])]

    rivals: list[str] = []
    for sibling_id in fam.get("confused_with", []):
        sibling = families().get(sibling_id)
        if not sibling:
            continue
        for phrase in (normalise(p) for p in sibling.get("canonical", [])):
            # A phrase that is also one of our own aliases is not evidence against us.
            # "data science" belongs to data_scientist but appears in data_analyst's
            # alias list too, and must not penalise either.
            if phrase and phrase not in intent.aliases:
                rivals.append(phrase)
    intent.rivals = rivals
    return intent


def is_noise_title(title: str) -> bool:
    """Titles that carry no information and should never rank."""
    t = normalise(title)
    return not t or len(t) < 3 or t in noise_titles()


@lru_cache(maxsize=1)
def _skill_vocabulary() -> list[tuple[str, str]]:
    """Every skill phrase in the taxonomy, as (normalised phrase, display form).

    Longest first, so "machine learning" is matched and consumed before "learning" would
    be, and "power bi" before "bi".
    """
    seen: dict[str, str] = {}
    for fam in taxonomy()["families"]:
        for skill in fam.get("skills", []):
            key = normalise(skill)
            if key and key not in seen:
                seen[key] = skill
    return sorted(seen.items(), key=lambda kv: -len(kv[0]))


def extract_skills(*texts: str, limit: int = 10) -> list[str]:
    """Pull known skills out of free text, without an LLM.

    This exists because the `skills` field was empty on 100% of the live corpus — it is
    populated by LLM enrichment, and with no API key configured that never ran. Since
    search matches against title, company, skills, tags and summary, and three of those
    were always blank, search was effectively title-only.

    Matching a fixed vocabulary is far narrower than what a model would extract, but it is
    free, instant, deterministic, and it cannot invent a skill the posting never mentions.
    """
    blob = normalise(" ".join(t for t in texts if t))
    if not blob:
        return []

    found: list[str] = []
    for key, display in _skill_vocabulary():
        if len(found) >= limit:
            break
        # Word-boundary match so "r" does not fire on every word containing the letter,
        # and "ai" does not match "email".
        if re.search(rf"(?<![a-z0-9]){re.escape(key)}(?![a-z0-9])", blob):
            found.append(display)
    return found

File: frontend/src/lib/taxonomy.ts
/**
 * Query understanding in the browser.
 *
 * Mirrors backend/jobradar/taxonomy.py and reads the same shared/role_taxonomy.json, so a
 * search typed into the box resolves to the same role family the pipeline used when it
 * built the corpus. tests/test_relevance.py asserts the two stay in step.
 */
import taxonomyJson from "@shared/role_taxonomy.json";

interface Family {
  id: string;
  label: string;
  canonical: string[];
  aliases: string[];
  skills: string[];
  confused_with: string[];
}

const taxonomy = taxonomyJson as unknown as {
  families: Family[];
  noise_titles: string[];
  seniority_order: string[];
  seniority_terms: Record<string, string[]>;
};

const FAMILIES = new Map(taxonomy.families.map((f) => [f.id, f]));
const NOISE = new Set(taxonomy.noise_titles.map((t) => t.toLowerCase()));
const SENIORITY_ORDER = taxonomy.seniority_order ?? [];

// Abbreviation expansions. The word boundary must come before the optional dot, or "Sr."
// normalises to "senior." and stops matching anything.
const EXPANSIONS: [RegExp, string][] = [
  [/\bsr\b\.?/g, "senior"],
  [/\bjr\b\.?/g, "junior"],
  [/\bmgr\b\.?/g, "manager"],
  [/\beng\b\.?/g, "engineer"],
  [/\bdev\b\.?/g, "developer"],
  [/\bml\b/g, "machine learning"],
  [/\bai\b/g, "artificial intelligence"],
  [/\bbi\b/g, "business intelligence"],
  [/\bds\b/g, "data science"],
  [/\bswe\b/g, "software engineer"],
  [/\bsde\b/g, "software engineer"],
  [/\bpm\b/g, "product manager"],
];

export function normalise(text: string): string {
  let t = ` ${text.toLowerCase()} `;
  t = t.replace(/[^a-z0-9+#/&.\- ]+/g, " ");
  for (const [re, repl] of EXPANSIONS) t = t.replace(re, repl);
  return t.replace(/\s+/g, " ").trim();
}

export interface Intent {
  raw: string;
  normalised: string;
  terms: string[];
  familyId: string;
  aliases: string[];
  skills: string[];
  rivals: string[];
  /** The experience level the query asked for, "" when it asked for none. */
  seniority: string;
}

// (normalised phrase, level), longest first so "entry level" beats "entry". Roman-numeral
// and single-letter grade markers are dropped: they are title suffixes, not something
// anybody types, and "i" would fire on every query.
const SENIORITY_PHRASES: [string, string][] = Object.entries(taxonomy.seniority_terms ?? {})
  .flatMap(([level, terms]) =>
    terms.filter((t) => t.length > 2).map((t) => [normalise(t), level] as [string, string]),
  )
  .sort((a, b) => b[0].length - a[0].length);

/**
 * The experience level a query asks for.
 *
 * Without this, "entry level data scientist", "junior data scientist" and "data scientist"
 * were the same search — all three returned an identical 27 results led by Sr and
 * Principal roles, because the seniority word matched no title and was carried along as
 * dead weight in the term list.
 */
export function detectQuerySeniority(queryNorm: string): string {
  for (const [phrase, level] of SENIORITY_PHRASES) {
    const re = new RegExp(`(?<![a-z0-9])${phrase.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}(?![a-z0-9])`);
    if (re.test(queryNorm)) return level;
  }
  return "";
}

/** How many rungs apart two levels are, or 0 when either is unknown. */
export function seniorityDistance(asked: string, offered: string): number {
  const a = SENIORITY_ORDER.indexOf(asked);
  const b = SENIORITY_ORDER.indexOf(offered);
  if (a < 0 || b < 0) return 0;
  return Math.abs(a - b);
}

function allPhrases(f: Family): string[] {
  return [...f.canonical, ...f.aliases].map(normalise);
}

export function detectFamily(queryNorm: string): string {
  // Longest phrase wins, so "machine learning engineer" beats "machine learning".
  let bestId = "";
  let bestLen = 0;
  for (const f of taxonomy.families) {
    for (const phrase of allPhrases(f)) {
      if (phrase && queryNorm.includes(phrase) && phrase.length > bestLen) {
        bestId = f.id;
        bestLen = phrase.length;
      }
    }
  }
  return bestId;
}

export function understand(query: string): Intent {
  const norm = normalise(query);
  const intent: Intent = {
    raw: query,
    normalised: norm,
    terms: norm.split(" ").filter((t) => t.length > 1),
    familyId: "",
    aliases: [],
    skills: [],
    rivals: [],
    seniority: detectQuerySeniority(norm),
  };

  const fid = detectFamily(norm);
  if (!fid) return intent;

  const fam = FAMILIES.get(fid)!;
  intent.familyId = fid;
  intent.aliases = allPhrases(fam);
  intent.skills = fam.skills.map(normalise);

  for (const siblingId of fam.confused_with) {
    const sibling = FAMILIES.get(siblingId);
    if (!sibling) continue;
    for (const phrase of sibling.canonical.map(normalise)) {
      // A phrase that is also one of our own aliases is not evidence against us.
      if (phrase && !intent.aliases.includes(phrase)) intent.rivals.push(phrase);
    }
  }
  return intent;
}

export function isNoiseTitle(title: string): boolean {
  const t = normalise(title);
  return !t || t.length < 3 || NOISE.has(t);
}
